Polkawatch integration - Validator Decentralization Modal View

rvalle opened the pull request:
This PR Integrates Polkawatch analytics with the staking dashboard and adds Validator Decentralization Visualization.

- The Polkawatch API has been created as a Provider
- A doughnut chart has been created with chart-js to display decentralization metrics.
- Automatic gradients are used based on theme colors to display charts with multiple series, by using chrome-js
- Decentralization charts may reduce labels, if they hold too much data, by using ellipsize
- An menu option to View decentralization has been added on ValidatorList, 
- Validators with no analytics (i.e., no rewards generated) or Networks without Decentralization analytics are handled properly with warning messages..
- Labels have been localized to Chinese. 
- Polkawatch has been registered as a plugin and can be Disabled/Enabled
- Responsiveness of the UI has been addressed
- Colors have been tested on Light/Dark, Kusama and Polkawatch.
- Help items are added under validator key.

After this PR adding other visualizations will be trivial.

This is how it looks: 
![Screenshot 2023-08-28 at 15-47-52 Polkadot Staking Dashboard Polkadot Staking (DOT)](https://github.com/paritytech/polkadot-staking-dashboard/assets/412837/f126243a-4ac6-43a8-9b5b-107b1b573e6e)

rvalle commented:
@rossbulat looks like a lot has changed in main recently. I am not sure this is merging... let me review.

rossbulat commented on src/contexts/Polkawatch/index.tsx:
```diff
@@ -0,0 +1,76 @@
+// Copyright 2023 @paritytech/polkadot-staking-dashboard authors & contributors
```
Perhaps this `Polkawatch` folder can sit inside the `Plugins` folder, so we have them all under one umbrella. So `src/contexts/Plugins/Polkawatch/index.tsx`.

rvalle replied:
Implemented

rossbulat commented on src/contexts/Polkawatch/index.tsx:
```diff
@@ -0,0 +1,76 @@
+ * in context. Also returns information about whether there exist decentralization analytics for the Network.
+ */
+
+const POLKAWATCH_API_VERSION = 'v2';
+const POLKAWATCH_NETWORKS = ['polkadot', 'kusama'];
```
To maintain conventions these values can go inside a `defaults.ts` file in this same folder. Check out the other contexts to check out the convention.

rvalle replied:
applied

rossbulat commented on src/contexts/Polkawatch/index.tsx:
```diff
@@ -0,0 +1,76 @@
+ * The provider will return an API and also information about whether the selected network has decentralization
+ * analytics support.
+ */
+interface PolkawatchState {
+  pwApi: PolkawatchApi;
+  networkSupported: boolean;
+}
```
This can go in a `types.ts` file alongside this file. Yet another convention of the current contexts structure :)

rvalle replied:
implemented

rossbulat commented on src/contexts/Polkawatch/index.tsx:
```diff
@@ -0,0 +1,76 @@
+ * @constructor
+ */
+const apiConfiguration = (
+  name: NetworkName = 'polkadot',
```
This initial network name may be different if the user has a different initial network set in local storage. 

```suggestion

  name: NetworkName = localStorageOrDefault('network', 'polkadot');
```

rvalle replied:
that is a good one, would have messed up things, potentially. implemented.

rossbulat commented on src/library/Graphs/DistributionDoughnut.tsx:
```diff
@@ -0,0 +1,95 @@
+// Copyright 2023 @paritytech/polkadot-staking-dashboard authors & contributors
```
Striving for a shorter name: `GeoDonut`?

rvalle replied:
Sounds good to me... Applying the change.

rossbulat commented on package.json:
```diff
@@ -34,13 +34,16 @@
     "chart.js": "^4.4.0",
+    "chroma-js": "^2.4.2",
     "date-fns": "^2.29.3",
+    "ellipsize": "^0.5.1",
```
We can remove this dependency and use `ellipsisFn` from `@polkadot-cloud/utils` courtesy of @wirednkod. 

Full code and docs: https://github.com/paritytech/polkadot-cloud/blob/main/packages/utils/lib/index.ts#L466-L470

rvalle replied:
@rossbulat I think ellipsisFN is broken.

Testing ellipsisFN('Canada',30,'right') and am getting "..." as result.

Same for 'United States' is returning 'Uni..."

I am going to apply the change anyway, and then look into this function. I don't know if there are unit tests, but probably would be best to just add some strange cases.

I will look into it later, unless you want to have a look first.

wirednkod replied:
Thank you @rvalle 
Based on our documentation, the 2nd argument defines `The amount of characters that the ellipsis will be`, meaning how many characters will appear, before the ellipsis is applied`.

Please note that there were some changes on the "left"/"right" (now "start"/"end") on latest release of `@polkadot-cloud/utils` v0.0.14 (2 days ago).

Though I did identified a bug thanks to you and I created: https://github.com/paritytech/polkadot-cloud/pull/598 which will result to v0.0.15.

Based on that latest release, when an amount of characters larger than the length of the string is given (while `start` or `end`), then this defaults to  `str.length - 3` + '...': 
Your code `ellipsisFN('Canada',30,'end')` which should result to: `Can...` and `ellipsisFN('United states',30,'end')` which should result to: `United sta...` 

If though you submit `ellipsisFN('United states',8 ,'end')` you should get: `United s...` while

rvalle replied:
I merged main, with the .15 

@wirednkod @rossbulat I think the there is a case that ellipsisFN is not considering.

When the provided string size is smaller than N, then there is no ellipsis required and the string should be returned intact.

`
ellipsisFN("Malta", 13, "end") = "Ma..."
` 

However "Malta" is only 5, which is less than 13, and it should not need to be "ellipsized" anyway.  I suggest in that case the string is returned unchanged. I could do that check from outside but I believe it belongs in the function.

The thing is that the semantics of the library I used before and this one are different: previously you specify the number of characters you have space for, and the library will adjust the string as/if required.

My guess is that our ellipsisFN was traditionally used for Account Numbers, which are long and fixed-length, and therefore variable strings as input were not an use case.

rvalle replied:
I have reviewed the ellipsisFN and I realize that is has been designed in a different way. Being used for long, constat size addresses the interface is designed to decide how you want the result to look, rather than using the UI space available.

Perhaps it is better that I deal with when to ellipsize/not ellipsize for now.

rvalle replied:
I have now fixed this on my side, without needed to touch ellipsizeFN.

wirednkod replied:
Thank you @rvalle. The underlying concept of the "ellipse" function transcends mere automation of an ellipsis, which can be accomplished straightforwardly with CSS. Instead, it aims to provide to the user an ellipsis function that grants control over the number of characters to display, allowing for manual determination, and positioning the ellipsis (whether at the start, center, or end of the string). Initially conceived for handling the lengthy addresses of an account, this functionality though can be extended to other character strings as well.

The implementation includes specific provisions to address scenarios, where the character count designated as "to show", exceeds the length of the string (typically defaults to 4, if I remember correctly). More details on the usage, are available within the documentation page of the polkadot-cloud project. If there is a case that is not covered , I am more than willing to incorporate it. Please do not hesitate to create an issue in the [polkadot-cloud](https://www.github.com/paritytech/polkadot-cloud/issues) repository if such a situation arises.

rossbulat commented on package.json:
```diff
@@ -60,6 +63,8 @@
   "devDependencies": {
     "@ledgerhq/logs": "^6.10.1",
+    "@types/chroma-js": "^2.4.0",
+    "@types/ellipsize": "^0.1.1",
```
```suggestion
```

rvalle replied:
applied change, pending debug of ellipsizeFN alternative

rossbulat replied:
Can this be removed now?

rossbulat commented on src/locale/cn/help.json:
```diff
@@ -332,6 +332,13 @@
           "在这种情况下, \"无\" 会显示在存储下的 \"为最低存款存储\"部分."
         ]
       ],
+      "rewardsByCountryAndNetwork": [
```
@TingALin is this all good?

TingALin replied:
"rewardsByCountryAndNetwork": [
        "按国家和网络划分的收益",
        [
          "显示来自不同国家和IP网络收益的预估百分比.",
          "Polkawatch分布式分析器是通过聚合验证人的IP位置的收益来计算的, 取样期为过去的60天."
        ]
      ],

rvalle replied:
thanks for the fix, applied.

rossbulat commented on src/overlay/index.tsx:
```diff
@@ -35,6 +35,7 @@ import { UpdateController } from '../modals/UpdateController';
 import { UpdatePayee } from '../modals/UpdatePayee';
 import { UpdateReserve } from '../modals/UpdateReserve';
 import { ValidatorMetrics } from '../modals/ValidatorMetrics';
+import { ValidatorDecentralization } from '../modals/ValidatorDecentralization';
```
Again striving for a shorter name :)

```suggestion
import { ValidatorGeo } from '../modals/ValidatorGeo';
```

rvalle replied:
I am happy to rename for making the code easier to read, shorter, etc. 

Geolocation is related to decentralization practically speaking.

However I have one reservation:

User facing I suggest we keep using the term "decentralization". For example in titles and help.

Polkawatch since the start is "decentralization analytics", and in many regards is a project designed to raise awareness among stakers about the need to "watch" the decentralization of our network. So, there is a lot of blog posts, documentation etc using this nomenclature. Dropping it completely would be counter productive.

rvalle replied:
This has been applied, with comments.

rossbulat reviewed (commented):
Excellent addition to the dashboard. 

Beyond my comments, we may have a light-weight version of the Donut available shortly as a Polkadot Cloud component. @wirednkod is exploring this, which will allow us to shrink the bundle size further. Since we are early in the release cycle I think it is fine to merge without this optimisation & replace when the component is ready too.

wirednkod commented:
> Excellent addition to the dashboard.
> 
> Beyond my comments, we may have a light-weight version of the Donut available shortly as a Polkadot Cloud component. @wirednkod is exploring this, which will allow us to shrink the bundle size further. Since we are early in the release cycle I think it is fine to merge without this optimisation & replace when the component is ready too.

This is now [ready and merged](https://github.com/paritytech/polkadot-cloud/pull/578). It exists under the `@polkadot-cloud/react` release `v1.52`

rvalle commented:
@rossbulat , @TingALin thanks for the comments, will review shortly.

The CN localization needs review.

rossbulat commented:
Thanks @rvalle! I'm on standby for anything I can assist with.

We have [a lightweight pie /  donut](https://github.com/paritytech/polkadot-staking-dashboard/pull/1445) PR almost ready to go thanks to @wirednkod - it looks like we'll be able to integrate it in this PR now.

rossbulat commented on src/locale/en/help.json:
```diff
@@ -344,6 +344,13 @@
+      "rewardsByCountryAndNetwork": [
+        "Rewards By Country And Network",
+        [
+          "Shows the estimated percentage of rewards that are generated from different Countries and IP Networks.",
```
```suggestion
          "Shows the estimated percentage of rewards that are generated from different countries and IP networks.",
```

rvalle replied:
typo fixed

rossbulat commented on src/locale/en/help.json:
```diff
@@ -344,6 +344,13 @@
+        "Rewards By Country And Network",
+        [
+          "Shows the estimated percentage of rewards that are generated from different Countries and IP Networks.",
+          "Polkawatch Decentralization analytics are computed by aggregating rewards by the IP Geolocation of the Validator. The sample period is the last 60 days."
```
```suggestion
          "Polkawatch decentralization analytics are computed by aggregating rewards by the IP Geolocation of the Validator. The sample period is the last 60 days."
```

rvalle replied:
applied.

rossbulat commented on src/modals/ValidatorDecentralization/index.tsx:
```diff
@@ -0,0 +1,154 @@
+    <>
+      <Title title={t('validatorDecentralization')} />
+      <div className="header">
+        <PolkadotIcon
+          dark={mode === 'dark'}
+          nocopy
+          address={address}
+          size={33}
+        />
```
As per my apology at Sub0 about the regular breaking changes, I recently updated this component to be named `Polkicon`, it disables copying by default, and the background color now refers to `--background-default`. So we can omit the `dark` and `nocopy` props.

The resulting JSX is as follows:

```suggestion
        <Polkicon
          address={address}
          size={33}
        />
```

rvalle replied:
this has been applied, no worries for refactoring code.

TingALin commented on src/modals/ValidatorDecentralization/index.tsx:
```diff
@@ -0,0 +1,154 @@
+              }}
+            >
+              <DistributionDoughnut
+                title="Rewards"
```
title={t('rewards')}

rvalle replied:
applied

TingALin commented on src/modals/ValidatorDecentralization/index.tsx:
```diff
@@ -0,0 +1,154 @@
+                }}
+              >
+                <DistributionDoughnut
+                  title="Rewards"
```
title={t('rewards')}

rossbulat replied:
Good spot!

rvalle replied:
applied

TingALin commented on src/locale/cn/modals.json:
```diff
@@ -59,6 +59,8 @@
     "controllerImported": "您必须导入您的Controller帐户才能签署此交易",
     "dashboardTips": "提示消息",
     "days": "天",
+    "decentralizationAnalyticsNotAvailable": "去中心化分析不可用",
```
"decentralizationAnalyticsNotAvailable": "无法使用该分布式分析器",
 "decentralizationAnalyticsNotSupported": "此网络不支持该分布式分析器",

rvalle replied:
applied.

TingALin commented on src/locale/cn/modals.json:
```diff
@@ -189,6 +192,7 @@
     "reserveText": "控制预存多少{{unit}}用于支付交易费用. 这笔金额是在您账户最低存款基础上增加的.",
     "resetLedgerAccounts": "重设Ledger帐户",
     "rewards": "收益",
+    "rewardsByCountryAndNetwork": "按国家和网络供应商收益 ",
```
"rewardsByCountryAndNetwork": "按国家/地区和网络供应商划分的收益",

rvalle replied:
applied

TingALin commented on src/locale/cn/modals.json:
```diff
@@ -249,6 +253,7 @@
     "updatePoolCommission": "更新提名池佣金设置",
     "updateWhoClaimRewards": "更改代表申领奖励地址",
     "updated": "己更新",
+    "validatorDecentralization": "验证器去中心化",
```
"validatorDecentralization": "验证人的分布式地理指标",

rvalle replied:
applied

TingALin commented on src/locale/cn/library.json:
```diff
@@ -171,6 +171,7 @@
     "validatingParachainBlocks": "验证平行链区块",
     "validatorCommission": "验证人佣金",
     "valueTooSmall": "值太小",
+    "viewDecentralization": "查看去中心化",
```
"viewDecentralization": "分布式指标",
"viewMetrics": "验证人指标",

rvalle replied:
applied.

TingALin commented:
> The CN localization needs review.

I have added some translation comments, the rest looks all good to me so far :)

rvalle commented:
@rossbulat, I think I have addressed all issues. Please review.

I need to check the ellipsis again once the .15 is merged in main. 

With regards to the chart, I think creating a common component is the right way to go. However note that Polkawatch visualizations can be very dense and complicated in some cases. I have made some notes in the @wirednkod PR.

Perhaps we could merge the GeoDonut  chart in a second PR, this PR is now big and difficult to maintain.  Besides I am happy to get involved with the common Donut. 

GeoDonut is not complete either. Works for now because validator decentralization is very simple, they are rarely in 2 or 3 countries. However other visualizations involve 10 entities easily. So this needs additional work anyway.

rvalle commented:
Hi @wirednkod @rossbulat 

This chart is quite dense in some decentralization use cases, validator decentralization, the first use case implemented is fairly simple. 

I attach an example from Polkawatch native site. 

![SampleComplexDonut](https://github.com/paritytech/polkadot-cloud/assets/412837/0f5ad414-f7af-4626-a0ed-8ada985331c1)

Variable length legend is very desirable, and the possibility to expand information with a popup also. The center display is not crucial.

rossbulat commented:
I believe @wirednkod fixed the `ellipsisFn`, if that can be used now I will go ahead and merge thereafter. We can use Chart.js for the charts until we are ready to migrate to the Cloud components.

rossbulat commented on src/library/Graphs/GeoDonut.tsx:
```diff
@@ -0,0 +1,112 @@
+};
+
+/**
+ * Depending on how much text we need to display in the legend it may be needed to cut the labels more or less.
+ * This function reuses polkadot ellipsize function to achieve that purpose when required.
+ * @param str the string to adjust
+ * @param labelTotal
+ */
+const ellipsize = (str: string, labelTotal: number) => {
+  // The maximum we can afford to present
+  const maxLength = labelTotal > 30 ? 10 : 20;
+
+  // We may not need to ellipsize in some cases
+  if (str.length <= maxLength) return str;
+
+  // Note that this function will add 3 characters '...'
+  return ellipsisFn(str, maxLength - 3, 'end');
+};
```
Was this a temporary workaround while `ellipsisFn` was being fixed up?
```suggestion
```

rossbulat commented on src/library/Graphs/GeoDonut.tsx:
```diff
@@ -0,0 +1,112 @@
+  );
+};
+
+/**
```
```suggestion
```

rvalle commented:
@rossbulat I also agree with hidding the menu item when the plugin is disabled. I think it is best.
